whetstone.py

```diff
@@ -32,22 +59,482 @@ def _authorize(self):
+            "private",
+            "signatureRequired",
+            "course",
+            "date",
+            "grade",
+            "school",
+            "title",
+            "type",
+            "creator",
+            "district",
+            "created",
+            "lastModified",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {
+            "Meetings": [],
+            "MeetingObservations": [],
+            "MeetingParticipants": [],
+            "MeetingAdditionalFields": [],
+        }
+        for record in records:
+            record_id = record.get("_id")
+            record["creator"] = record.get("creator").get("_id")
+            meeting = {k: v for (k, v) in record.items() if k in self.columns}
+            models["Meetings"].append(meeting)
+
+            observations = record.get("observations")
+            if observations:
+                observations = [
+                    dict(meeting=record_id, observation=observation)
+                    for observation in observations
+                ]
+                models["MeetingObservations"].extend(observations)
+
+            participants = record.get("participants")
+            if participants:
+                participants = [dict(item, meeting=record_id) for item in participants]
+                models["MeetingParticipants"].extend(participants)
+
+            additional_fields = record.get("additionalFields")
+            if additional_fields:
+                additional_fields = [
+                    dict(item, meeting=record_id, content=str(item.get("content")))
+                    for item in additional_fields
+                ]
+
+                models["MeetingAdditionalFields"].extend(additional_fields)
+
+        return models
+
+
+class Observations(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "observedAt",
+            "observedUntil",
+            "firstPublished",
+            "lastPublished",
+            "viewedByTeacher",
+            "isPublished",
+            "archivedAt",
+            "requireSignature",
+            "locked",
+            "isPrivate",
+            "signed",
+            "observer",
+            "rubric",
+            "teacher",
+            "district",
+            "observationType",
+            "observationModule",
+            "observationtag1",
+            "observationtag2",
+            "observationtag3",
+            "created",
+            "lastModified",
+            "quickHits",
+            "score",
+            "scoreAveragedByStrand",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {
+            "Observations": [],
+            "ObservationScores": [],
+            "ObservationMagicNotes": [],
+        }
+        for record in records:
+            record_id = record.get("_id")
+            observation = {
+                k: v["_id"] if type(v) is dict else v
```

Comment:
Is this if logic here because the item values types are inconsistent?

Reply:
I'd love to know what's happening here as well. What is the structure of the data the API is returning?

Reply:
@susanxiong @zkagin The observations endpoint has a number of nested dictionaries for user objects that include additional info like the string representation of the user's name, but for relational mapping we only need the id. Rather than explicitly add these one by one, I opted for this approach to handle them across the board.